### PR TITLE
feat(gatsby): plugin option validation

### PR DESCRIPTION
**Pull request by mxstbr**
This implements plugin option validation if a plugin has a `gatsbyNode.pluginOptionsSchema` defined. (see #27166 for the RFC for the `pluginOptionsSchema` API)

I think we should disable this by default until we have documentation for it, added some more schemas to core plugins and (maybe) have tested it with a few alpha/beta testers! So, this PR adds a feature flag in the form of the `GATSBY_EXPERIMENTAL_PLUGIN_OPTION_VALIDATION` environment variable. 

I added the Google Analytics option schema for testing. Here is an example of what a configuration error looks like in the terminal with these incorrect options:

```JS
{
  resolve: `gatsby-plugin-google-analytics`,
  options: {
    anonymize: `not a boolean`
  }
}
```

Screenshot 2020-10-05 at 11 48 27

[ch15946]

### CODE REVIEW

**Review by laurieontech — COMMENTED** [review]
Liking the structured errors!

File: packages/gatsby/src/bootstrap/load-plugins/validate.ts
Comment (ascorbic) [inline packages/gatsby/src/bootstrap/load-plugins/validate.ts:204]: No need for optional chain here, as you have the falsy check right before

File: packages/gatsby/src/bootstrap/load-plugins/validate.ts
Comment (ascorbic) [inline packages/gatsby/src/bootstrap/load-plugins/validate.ts:228]: What happens if the plugin defines their own schema for `plugins`?

File: packages/gatsby/src/bootstrap/load-plugins/validate.ts
Comment (ascorbic) [inline packages/gatsby/src/bootstrap/load-plugins/validate.ts:236]: Does the structured error need implementing in cloud before this merges?

**Review by ascorbic — COMMENTED** [review]
Are we able to add a definition for `GatsbyNode["pluginOptionsSchema"]` to gatsby/index.d.ts? I'm assuming we'll need to add our own type for Joi ones we extend it? Do you have any idea how that works in general?

**Comment by ascorbic** [comment]
Looking good!

File: packages/gatsby/src/bootstrap/load-plugins/validate.ts
Comment (mxstbr) [reply packages/gatsby/src/bootstrap/load-plugins/validate.ts:232]: Great question! Clarified internally that it does not 👍

File: packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.ts
Comment (wardpeet) [inline packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.ts:263]: Do we want inlineSnapshot?

File: packages/gatsby-plugin-google-analytics/src/gatsby-node.js
Comment (wardpeet) [inline packages/gatsby-plugin-google-analytics/src/gatsby-node.js:39]: can we move this out of the PR as it's not really part of this change.

File: packages/gatsby-plugin-utils/src/joi.ts
Comment (wardpeet) [inline packages/gatsby-plugin-utils/src/joi.ts:24]: I might have forgotten but I'm still not sure why we want this one.

**Review by wardpeet — COMMENTED** [review]
Some small nits and questions

File: packages/gatsby-plugin-utils/src/joi.ts
Comment (mxstbr) [reply packages/gatsby-plugin-utils/src/joi.ts:24]: This let's plugin authors mark that a value entered during (a potential) `gatsby plugin add` command should be stored in the `.env` file instead of being written to the gatsby-config.js directly. E.g.:

```JS
// gatsby-plugin-source-contentful/gatsby-node.js
export const pluginOptionsSchema = ({ Joi }) => Joi.object({
  accessToken: Joi.string().required().dotenv("CONTENTFUL_ACCESS_TOKEN")
})
```

Then, when we prompt the user to enter their access token we will 1) mask the input in the CLI and 2) write the value to the .env file and `process.env.CONTENTFUL_ACCESS_TOKEN` to the gatsby-config.js 👍

File: packages/gatsby-plugin-google-analytics/src/gatsby-node.js
Comment (mxstbr) [reply packages/gatsby-plugin-google-analytics/src/gatsby-node.js:39]: I use it in the tests.

File: packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.ts
Comment (mxstbr) [reply packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.ts:263]: Personal preference — I like inline snapshot a lot more because with `toMatchSnapshot` the result is hidden in some folder and file that nobody ever looks at / can understand! If you prefer to move it out, I'd be happy to use `.tomatchSnapshot` though!

File: packages/gatsby-plugin-utils/src/joi.ts
Comment (wardpeet) [reply packages/gatsby-plugin-utils/src/joi.ts:24]: It's still something I'm not 100% convinced about. The feature sounds really cool, but I fear that env vars can be similar and bring conflicts.

Can we strip it out until we have the need for it/more discussions?

File: packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.ts
Comment (wardpeet) [reply packages/gatsby/src/bootstrap/load-plugins/__tests__/load-plugins.ts:263]: Honestly, I don't care but we usually use matchSnapshot, that's why I asked

File: packages/gatsby-plugin-utils/src/joi.ts
Comment (laurieontech) [reply packages/gatsby-plugin-utils/src/joi.ts:24]: Isn't this the main way we can ensure core -> cloud compatibility? @wardpeet can you explain more about what conflicts you anticipate?

Comment (wardpeet) [reply packages/gatsby-plugin-utils/src/joi.ts:24]: I might not 100% understand it so let's say that I have plugin A that has `dotenv('HOST')` and a plugin B `dotenv('HOST')` how would this play out? Also what if for some reason I rather have CONTENTFUL_DELIVERY_TOKEN  instead of CONTENTFUL_ACCESS_TOKEN?

Comment (laurieontech) [reply packages/gatsby-plugin-utils/src/joi.ts:24]: So this isn't for the user to understand, but rather the plugin author. We could solve the duplicate problem by concatenating the plugin name and the option name. So `GATSBY_PLUGIN_CONTENTFUL_HOST`. 

The core <-> compatibility is the idea that you'll always be in a valid state and you'd be ready to set up with cloud since it leverages process.env to hook into CMS keys and the like.

**Comment by MichaelDeBoey** [comment]
@mxstbr Will it also be possible to validate that we don't pass any options anymore?
For instance in a previous version of a plugin you could pass some options, but they have been removed in a new major version of the plugin and the plugin author wants to notify the user about that.

File: packages/gatsby-plugin-utils/src/joi.ts
Comment (wardpeet) [reply packages/gatsby-plugin-utils/src/joi.ts:24]: That's why I want to hold off on adding this feature cause it needs more thinking.

**Comment by mxstbr** [comment]
> Will it also be possible to validate that we don't pass any options anymore?
For instance in a previous version of a plugin you could pass some options, but they have been removed in a new major version of the plugin and the plugin author wants to notify the user about that.

This'll happen automatically! When you publish a new version with a changed options schema the user will get an error after upgrading telling them that they are using options that don't exist 👍

File: packages/gatsby-plugin-utils/src/joi.ts
Comment (mxstbr) [reply packages/gatsby-plugin-utils/src/joi.ts]: I have disabled (i.e. commented out) this extension for now until we decide on a fixed API for it to unblock this PR 👍

**Review by wardpeet — APPROVED** [review]
Looks awesome! 👍 Let's ship and test it out.

**Comment by openscript** [comment]
I like the idea! .. but it brought some troubles to me. I'm using `gatsby-plugin-ts-config` to write typed configuration files. Now I'm forced to define a `pluginOptionsSchema` property, when I configure my `GatsbyNode`.

The comment says `Currently experimental and not enabled by default.` so why is it not typed as optional?

    pluginOptionsSchema?(args: PluginOptionsSchemaArgs): ObjectSchema

**Comment by mxstbr** [comment]
Great catch @openscript, it should definitely be optional! Would you mind opening a new PR with that change? :pray:

File: packages/gatsby-plugin-google-analytics/src/gatsby-node.js
Comment (MichaelDeBoey) [inline packages/gatsby-plugin-google-analytics/src/gatsby-node.js:7]: @mxstbr Shouldn't we have kept `onPreInit` in case `GATSBY_EXPERIMENTAL_PLUGIN_OPTION_VALIDATION` wasn't `true`? 🤔

If people update `gatsby-plugin-google-analytics` but don't update `gatsby`, they won't get any warning anymore.

Comment (mxstbr) [reply packages/gatsby-plugin-google-analytics/src/gatsby-node.js:7]: True, good catch! PR welcome 👍

Comment (MichaelDeBoey) [reply packages/gatsby-plugin-google-analytics/src/gatsby-node.js:7]: On it boss! 😉

Comment (MichaelDeBoey) [reply packages/gatsby-plugin-google-analytics/src/gatsby-node.js:7]: Here you go: https://github.com/gatsbyjs/gatsby/pull/27495 🙂

**Comment by kamranayub** [comment]
Hey team, working on a Gatsby plugins course and I wanted to feature this plugin options schema support but it _appears_ that it isn't working for a local plugin. 🤔 

I still have to dig into this PR/the code and see if I can confirm that's the issue. If so I'll open a new ticket and try and contribute a fix since I'd love to showcase this (non-local plugins are introduced later in the course and the local plugin in my demo is a good use case for options schema).

**Comment by KyleAMathews** [comment]
That'd be sweet!

**Comment by kamranayub** [comment]
Yes, looks like it does fail with a small test case so I will open an issue.

image

```js
it(`validates local plugin schemas`, async () => {
      await loadPlugins({
        plugins: [
          {
            resolve: require.resolve('./__fixtures__/local-plugin'),
            options: {
              optionalString: 1234
            },
          },
        ],
      })

      expect(reporter.error as jest.Mock).toHaveBeenCalledTimes(1)
      expect(mockProcessExit).toHaveBeenCalledWith(1)
    })
```

**Comment by kamranayub** [comment]
For whomever would like to review, I have a proposed fix: https://github.com/gatsbyjs/gatsby/pull/29787

**Pull request merged**
